packages/assets-controllers/src/MultichainAssetsController/MultichainAssetsController.ts

```diff
@@ -0,0 +1,493 @@
+    });
+  }
+
+  /**
+   * Creates a mapping of CAIP-2 Chain ID to Asset Snaps.
+   *
+   * @returns A mapping of CAIP-2 Chain ID to Asset Snaps.
+   */
+  #getAssetSnaps(): Record<CaipChainId, Snap[]> {
+    const snaps: Record<CaipChainId, Snap[]> = {};
+    const allSnaps = this.#getAllSnaps();
+    const allPermissions = allSnaps.map((snap) =>
+      this.#getSnapsPermissions(snap.id),
+    );
+    // start mock
+    allPermissions.forEach((singlePermission) => {
+      (singlePermission as unknown as AssetEndowment) = {
+        ...singlePermission,
+        'endowment:assets': {
+          scopes: ['bip122:000000000019d6689c085ae165831e93'],
+        },
+      };
+    });
+    (allPermissions[0] as unknown as AssetEndowment) = {
+      ...allPermissions[0],
+      'endowment:assets': {
+        scopes: ['solana:EtWTRABZaYq6iMfeYKouRu166VU2xqa1'],
+      },
+    };
+    // end mock
+
+    for (const [index, permission] of allPermissions.entries() as unknown as [
+      number,
+      AssetEndowment,
+    ][]) {
+      const scopes = permission['endowment:assets']?.scopes;
+      if (!scopes) {
+        continue;
+      }
+      for (const scope of scopes) {
+        if (!snaps[scope]) {
+          snaps[scope] = [];
+        }
+        snaps[scope].push(allSnaps[index]);
+      }
+    }
+    return snaps;
+  }
+
+  /**
+   * Returns the first asset snap for the given scope
+   *
+   * @param scope - The scope to get the asset snap for
+   * @returns The asset snap for the given scope
+   */
+  #getAssetSnapFor(scope: CaipChainId): Snap | undefined {
+    const allSnaps = this.#snaps[scope];
+    // Pick only the first one, we ignore the other Snaps if there are multiple candidates for now.
+    return allSnaps?.[0]; // Will be undefined if there's no Snaps candidate for this scope.
+  }
+
+  /**
+   * Returns all the asset snaps
+   *
+   * @returns All the asset snaps
+   */
+  #getAllSnaps(): Snap[] {
+    return this.messagingSystem.call('SnapController:getAll') as Snap[];
+  }
+
+  /**
+   * Returns the permissions for the given origin
+   *
+   * @param origin - The origin to get the permissions for
+   * @returns The permissions for the given origin
+   */
+  #getSnapsPermissions(
+    origin: string,
+  ): SubjectPermissions<PermissionConstraint> {
+    return this.messagingSystem.call(
+      'PermissionController:getPermissions',
+      origin,
+    ) as SubjectPermissions<PermissionConstraint>;
+  }
+
+  /**
+   * Returns the metadata for the given assets
+   *
+   * @param assets - The assets to get metadata for
+   * @param snapId - The snap ID to get metadata from
+   * @returns The metadata for the assets
+   */
+  async #getMetadata(
+    assets: CaipAssetType[],
+    snapId: string,
+  ): Promise<Record<CaipAssetType, AssetMetadata>> {
+    return (await this.messagingSystem.call('SnapController:handleRequest', {
+      snapId: snapId as SnapId,
+      origin: 'metamask',
+      handler: HandlerType.OnRpcRequest,
```

Comment:
```suggestion
      handler: HandlerType.OnAssetsLookup,
```

Reply:
I think the OnAssetsLookup is available on [8.9.0](https://github.com/MetaMask/snaps/compare/@metamask/snaps-utils@8.8.0...@metamask/snaps-utils@8.9.0);

I wonder if its not better to have a separate PR to bump the following packages:
`@metamask/snaps-utils@8.9.0`
`@metamask/snaps-sdk@6.16.0`

Trying it locally looks like there could be other packages that get affected by this upgrade

Reply:
Updated Handler type after bumping snaps-utils here [8d90188](https://github.com/MetaMask/core/pull/5138/commits/8d9018846bc3abd89f1196d244ebd890e226c2cc) 🙏